liblepton/scheme/conf/upgrade.scm

```diff
@@ -0,0 +1,233 @@
+( let
+  (
+  ( cfg    #f )
+  ( fname  #f )
+  ( keys  '() )
+  ( data  '() )
+  )
+
+  ;
+  ; read legacy config file (geda*.conf):
+  ;
+
+  ( config-set-legacy-mode! #t ) ; use geda*.conf files
+
+  ( set! cfg ( get-cfg-ctx what ) )
+  ( unless cfg
+    ( throw 'ctx "Cannot get config context (input)" )
+  )
+
+  ( set! fname (config-filename cfg) )
+  ( unless ( and fname (access? fname F_OK) )
+    ( throw 'infile "Input config file does not exist:" fname )
+  )
+
+  ( format #t "ii: INPUT: ~a~%" fname )
+  ( format #t "~%" )
+
+  ( catch #t
+    ( lambda()
+      ( config-load! cfg )
+      ( set! keys (conf-keys-list) )
+      ( set! data
+        ( read-old cfg keys #:report-absent-keys report-absent-keys )
+      )
+    )
+    ( lambda( ex . args )
+      ( throw 'load "Cannot read config:" ex args )
+    )
+  )
+
+
+  ;
+  ; write new config file (lepton*.conf):
+  ;
+
+  ( format #t "~%" )
+
+  ( config-set-legacy-mode! #f ) ; use lepton*.conf files
+
+  ( set! cfg ( get-cfg-ctx what ) )
+  ( unless cfg
+    ( throw 'ctx "Cannot get config context (output)" )
+  )
+
+  ( set! fname (config-filename cfg) )
+  ( when ( and fname (access? fname F_OK) (not overwrite) )
+    ( throw 'outfile "Output config file already exists:" fname )
+  )
+
+  ( format #t "ii: OUTPUT: ~a~%" fname )
+
+  ( catch #t
+    ( lambda()
+      ( write-new cfg data )
+      ( config-save! cfg )
+    )
+    ( lambda( ex . args )
+      ( throw 'save "Cannot write config:" ex args )
+    )
+  )
+
+  ; return:
+  fname
+
+) ; let
+) ; convert()
+
+
+
+; get configuration context
+;
+; [what]: symbol: 'local, 'user or 'system
+; {ret}:  requested configuration context or #f on failure
+;
+( define ( get-cfg-ctx what )
+  ; return:
+  ( cond
+    (  ( eq? what 'local  )  ( path-config-context (getcwd) )  )
+    (  ( eq? what 'user   )  ( user-config-context )           )
+    (  ( eq? what 'system )  ( system-config-context )         )
+    (  else #f  )
+  )
+)
+
+
+
+; read legacy configuration
+;
+; [cfg]:                 configuration context to read from
+; [keys]:                cfg keys list as defined in (conf keys) module
+; [report-absent-keys]:  report absent keys
+; {ret}:                 list of entries: ( grp-new key-new val )
+;
+( define* ( read-old cfg keys #:key (report-absent-keys #f) )
+( let
+  (
+  ( grp-old  #f )
+  ( key-old  #f )
+  ( grp-new  #f )
+  ( key-new  #f )
+  ( pfn      #f )
+  ( val      #f )
+  ( res     '() )
+  )
+
+  ( define ( parse-entry e )
+    ( set! pfn     (list-ref e 0) )
+    ( set! grp-old (list-ref e 2) )
+    ( set! key-old (list-ref e 3) )
+    ( set! grp-new (list-ref e 4) )
+    ( set! key-new (list-ref e 5) )
+  )
+
+  ( define ( add-to-res g k v ) ; group, key, value
+    ( set! res ( cons (list g k v) res ) )
+  )
+
+  ( for-each
+  ( lambda( entry )
+
+    ( parse-entry entry )
+
+    ( catch #t
+      ( lambda()
+        ( set! val ( pfn cfg grp-old key-old ) ) ; pfn() throws
```

Comment:
One note for future. In Lisp world nobody promises that a sequence of procedures will be evaluated in order. While `guile` docs state it does so, things can be changed in future. Or, say, another Scheme could be used (a few years ago, TinyScheme has been suggested by one of the geda devs). Statements could be processed in parallel then. That is why local variable definition, say, by `let` is better than `set!`. Thus, the functional style of programming, when all procedure arguments do not depend upon others, is better. This is why, AIUI, the folded definitions are used so often. Results of low level functions are directly passed to top level ones. Lisp is not C, please remember this ;-)

Reply:
Using language that cannot provide basic code execution guaranties in any serious project would be disastrous.